### PR TITLE
tests: pin DNSBL probe/resolver parity; make the tick suites hermetic

### LINKED ISSUE #1666: TickEntrypointTest::testLogMaintenanceRunsEvenWhenFeedCronNotDue is order-dependent flaky on full-suite runs
## Summary

`TickEntrypointTest::testLogMaintenanceRunsEvenWhenFeedCronNotDue` is order-dependent: on full `vendor/bin/phpunit` runs it intermittently fails, then passes on isolated or repeated runs. Surfaced during independent verification of PR #1664 (issue #1652) — the failure is in the tick/ledger/log-trim path and is **unrelated** to that PR's diff (which touches only `www/index.php` + `LogFormatConsumersTest.php`), confirming it is pre-existing, not a regression introduced there.

## Evidence

- Full-suite run: 1 failure in `TickEntrypointTest::testLogMaintenanceRunsEvenWhenFeedCronNotDue`.
- Immediate re-runs (×2) of the full suite: clean (`Tests: 3916 ... OK`).
- Isolated `--filter` run of the test: passes.
- PR #1664 diff does not touch the tick/ledger code, so the flake is independent of that change.

## Why it matters

An order-dependent test undermines suite trust and can produce spurious CI reds. Per `testing.md`, a flake is fixed by making the test synchronize on the real state/ordering it depends on — never by widening a timeout or by re-run masking.

## Action

Find the cross-test state leak (shared ledger/log-trim fixture, global, or filesystem artifact not reset between tests) that makes this test's outcome depend on execution order, and make its setup/teardown hermetic so it passes regardless of order. Add an assertion that fails if the ordering assumption regresses.

### LINKED ISSUE #1711: DNSBL save-time probe duplicates the resolver's shape/budget rules with no parity pinning
## Summary

The DNSBL save-time regex validator hand-duplicates the resolver's admission rules in Python-inside-PHP,
and almost none of that duplication is pinned. #1688 closed the *reported* gap (the 200-character length
cap) and added a parity test for that one constant. The **class** of defect — save accepts, resolver
silently drops — remains one edit away from recurring through any of the other duplicated rules.

## What is duplicated, and what is pinned

Probe: `pfb_dnsbl_regex_validation_errors()` in `src/usr/local/pkg/pfblockerng/pfblockerng_extra.inc`.
Runtime: `src/usr/local/pkg/pfblockerng/pfb_unbound.py`.

| Duplicated rule | Pinned today | Consequence if it drifts |
|---|---|---|
| Length constant `200` | **yes** — `test_save_time_probe_length_cap_matches_runtime_constant` (#1688), mutation-proven | test goes red, loud |
| Probe-side comparator `>` | **yes, effectively** — the parity test's tag regex does not match `>=`, plus 200-accept/201-reject tests | test goes red |
| **Runtime-side comparator `>`** (three drop sites) | **no** | a `>` → `>=` drift at load stays green everywhere while save accepts a 200-character pattern the resolver drops — the #1688 class, at width 1 |
| **The four catastrophic-shape regexes** | **no** — byte-diffed manually once during review, nothing keeps it true | the resolver broadens or adds a shape → save accepts, load silently drops. **Full #1688 recurrence; highest hazard of the set** |
| **Budget regexes + threshold `12`** | **no** — probe hardcodes `12`, runtime has `_REGEX_BUDGET_MAX = 12` | boundary patterns accepted at save, dropped at load |
| Measured-string derivation | **no** | probe and runtime measure different strings; divergence in both directions (a known instance is #1707) |
| Check ordering | probe-side yes; runtime-side unpinned but the composite verdict is order-independent | cosmetic |

## Why this is worth its own issue

Every unpinned row reproduces exactly the user-visible failure #1656 and #1688 set out to eliminate: the
administrator is told the save succeeded, and the resolver never loads the pattern. The duplication has
been unpinned since it was introduced in #1667; #1688 raised the pinned count from zero to one.

The shape-regex row is the one to worry about. Those four expressions are the always-on safety gate, so
they are the rules most likely to be tuned in future — and tuning them on the runtime side alone would
silently reopen the defect with no test failing anywhere.

## Two suggested fixes, cheapest first

**1. Extend the pinning technique that already works.** #1688's parity test greps the probe's tagged
`> 200` line out of the `.inc` and compares it against `pfb_unbound.REGEX_STATIC_LEN_CAP`. The same
approach extends mechanically to the rest: extract the probe's `shapes = (...)` tuple and its budget
regexes and threshold from the `.inc` source, and byte-compare them against
`_REGEX_NESTED_QUANTIFIER.pattern`, `_REGEX_ALTERNATION_OVERLAP.pattern`,
`_REGEX_ADJACENT_GROUP_QUANTIFIER.pattern`, `_REGEX_STACKED_BOUNDED_REPEAT.pattern`,
`_REGEX_UNBOUNDED_QUANTIFIER.pattern`, `_REGEX_ALTERNATION.pattern` and `_REGEX_BUDGET_MAX`. Roughly
twenty lines of pytest, and it fails loudly on any drift.

Also add the missing **runtime** boundary case: feed an exactly-200-character rule to
`_dnsbl_compile_regex_rules(..., static_cap=True)` and assert it is admitted. Today the runtime tests only
use a ~242-character pattern and assert `len > CAP`, so the `>` versus `>=` distinction is untested at
load.

**2. Or remove the duplication.** Extract the pure static predicates and constants — roughly forty lines,
none of which touch Unbound — into a small shipped module both sides import.

Note for whoever evaluates option 2: the obvious objection, that `pfb_unbound.py` cannot be imported
off-appliance because of `unboundmodule`, is **wrong** — that import is `TYPE_CHECKING`-only. The actual
blocker is that module-level code calls `log_info`, a global injected by pythonmod, so a naive import dies
with `NameError: name 'log_info' is not defined` at module scope. The test suite works around this by
copying names from `stubs/python/unboundmodule.py` onto `builtins`, and `stubs/` is deliberately not
shipped. So option 2 means a new shipped file (which needs the three-port plist wiring this repo has been
bitten by before) plus a decision about that injected-name contract — real work, not a refactor.

Option 1 is the pragmatic close; option 2 is the durable one.

## Provenance

Raised by the two independent reviews on PR #1709 (#1688's fix), which each re-derived the parity
manually and observed that nothing keeps it true. Pre-existing since #1667; not chargeable to #1688,
which strictly improved the situation.

🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

### PR DESCRIPTION
**Pull request by andrebrait**
Two correlated test-architecture fixes from the v4.0.0 milestone, one commit per issue.

Fixes #1711. Fixes #1666.

## Commit 1 — pin the DNSBL save-probe/resolver parity surface (#1711)

Option 1 from the issue (extend the #1688 pinning technique):

- Tag comments inside the save-time probe's nowdoc (`pfblockerng_extra.inc`) mirroring the existing `# REGEX_STATIC_LEN_CAP mirror (pfb_unbound.py)` convention: the four catastrophic-shape regexes, the two budget regexes, and the budget threshold.
- New parity tests in `tests/test_adr07_regex_safety.py`: tag-driven literal extraction with an exact-set-of-6 assertion (a missing or extra tag fails loudly, keeping the enumeration self-maintaining), `ast.literal_eval` byte-comparison against the live `pfb_unbound` compiled patterns, threshold parity, and a 3-site strict-`>` comparator scan of the runtime.
- The previously-unpinned runtime admission boundary: exactly-200-char pattern admitted / exactly-201 dropped through the real `_dnsbl_compile_regex_rules(..., static_cap=True)`, mirroring the PHP-side `DnsblRegexEntryErrorTest` boundary pair.
- Deliberately NOT pinned (intentional deltas per the issue): the probe-only control-char rule and the measured-string derivation.

Evidence: four executed mutation proofs (shape literal flip, threshold 12→13, tag removal, runtime `>`→`>=`) each drove the intended test red, then green after revert. The probe comment edit is inert: the PHP suite executes the real probe (`DnsblRegexEntryErrorTest`, 57 tests green).

## Commit 2 — make the tick suites hermetic (#1666)

Root cause of the order-dependent flake (evidence in the commit): `TickApplyReconciliationTest` seeded no due-ledger entries, so every tick in that class fired a real backgrounded `pfblockerng.php dcc` exec; the leaked shell matches `pfb_update_pass_running()`'s ps-scan regex, and `pfblockerng_tick()`'s tail gate then silently skips `pfb_log_mgmt()` — exactly the assertion `testLogMaintenanceRunsEvenWhenFeedCronNotDue` rides.

Test-only fix: seed future-dated ledgers + a recording-stub `$pfb['php']` in the emitter suites (TickApplyReconciliation, TickEntrypoint C/D, TickFeedPassDeferral rows), sandbox and restore the process-global `$pfb['log']`/`logdir`/`errlog`, `$GLOBALS['config']`, and `unbound_chroot_path` in `TickEntrypointTest`, restore the four `$pfb` path keys in `AlertsPfctlCheckedSitesTest`'s teardown, and add two loud tripwires: a no-dispatch assertion (`testTickDispatchesNothingWhenNoJobIsDue`) and a precondition `assertFalse(pfb_update_pass_running(), ...)` before the tick — both proven RED against the respective regressions and GREEN after (evidence in the verification record).

Order-robustness: `--filter Tick` ×3 and the full suite ×2, all green. No production file touched; no timeout widened (per `testing.md`, deadlines are never assertions).

🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

## Summary by CodeRabbit

* **Tests**
  * Expanded coverage for regex safety validation, including exact length-limit boundaries and consistency between validation stages.
  * Improved test isolation to prevent background tasks and temporary configuration from affecting other test runs.
  * Added verification that scheduled maintenance is skipped when no work is due and proceeds correctly when required.

* **Refactor**
  * Clarified internal validation logic without changing existing acceptance or rejection criteria.

### CODE REVIEW

File: tests/php/AlertsPfctlCheckedSitesTest.php [hunk 1]
```diff
@@ -99,6 +109,14 @@ protected function setUp(): void
 		putenv("PFB_TEST_RULES={$this->rulesPath}");
 		$this->shim = $this->writeShim();
 
+		// issue #1666: save the four $pfb path keys this test overrides below,
+		// before overriding them, so tearDown() can restore them once $this->tmp
+		// (which they all point inside) is deleted.
+		foreach (['dbdir', 'logdir', 'log', 'errlog'] as $k) {
+			$this->hadPfb[$k]   = array_key_exists($k, $GLOBALS['pfb'] ?? []);
+			$this->savedPfb[$k] = $GLOBALS['pfb'][$k] ?? NULL;
+		}
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/php/AlertsPfctlCheckedSitesTest.php]: Applied in 4c859e78 — the snapshot/restore now covers every `$pfb` key `setUp()` overrides (all 11: dbdir, logdir, log, errlog, pfctl, aliasdir, permitdir, ip_unlock, supptxt, supptxt_v6, ipconfig), enumerated from the setUp body itself, plus `$GLOBALS['config']`, preserving existed-vs-absent in both directions. Restoration proven by an executed three-file scratch run (sentinel seeder → full suite → checker asserting every sentinel byte-identical afterward): OK (41 tests, 194 assertions).

The independent adversarial review reached the same finding; this reply covers both threads.

🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

**Comment by andrebrait** [comment]
## Review audit — PR #1764

**Adversarial review:** full profile, effort xhigh, top tier (Fable) — chosen by size metric (>300 changed lines across 6 files). Verdict REQUEST-CHANGES → fixed → delta re-review (small tier, base `0ea037bc`) APPROVE (clean round). **CodeRabbit:** reviewed (CHILL profile), 1 actionable finding. **Snyk:** no scan on the head SHA — skipped scan, not a clean pass. **CI:** all required checks green on final head `4c859e78661e`.

### Findings ledger

1. **[blocking, adversarial]** `TickApplyReconciliationTest` no-dispatch tripwire race-blind (assert ran before the backgrounded exec's log write; 6/6 green on the regression) — **fixed@4c859e78**: deterministic per-job due-ledger snapshot/`assertSame` (dispatch marks its ledger synchronously before `pfblockerng_tick()` returns); mutation-proven red 5/5 with seeding removed, re-verified independently in the delta review.
2. **[major, CodeRabbit + adversarial nitpick — deduped]** `AlertsPfctlCheckedSitesTest` restored only 4 of 11 overridden `$pfb` keys and not `$GLOBALS['config']` — **fixed@4c859e78**: full snapshot/restore (existed-vs-absent preserved); restoration proven by an executed sentinel seeder→suite→checker run (41 tests OK).
3. **[nitpick, adversarial]** recorder stub's own argv transiently matches `pfb_update_pass_running()`'s ps regex — **fixed@4c859e78** (documented; unavoidable: production builds the argv). Residual nit from the closing round: the comment cites `testing.md` where `git.md`'s one-worktree-per-agent rule is the closer source — noted here, not worth a CI cycle.
4. **[outside-diff, adversarial + maintainer question]** exact-set-of-6 parity pin could not catch a NEW runtime shape regex — **fixed@4c859e78**: the pinned set is now extracted from `_regex_is_catastrophic_shape`'s own source, so a seventh runtime shape without a probe mirror goes red; mutation-proven with a scratch `_REGEX_FAKE`.

Option 2 of #1711 (deduplicate probe/resolver into a shared module, retiring these pins) is tracked as #1765.

🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.
